### PR TITLE
JIT ARM64-SVE: Add AW_2A to AZ_2A, BM_1A, BN_1A

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Part of #94549. Adds the following encodings:
* SVE_AW_2A
* SVE_AX_1A
* SVE_AY_2A
* SVE_AZ_2A
* SVE_BM_1A
* SVE_BN_1A

cstool output:
```
xar   z0.b, z0.b, z1.b, #1
xar   z2.b, z2.b, z3.b, #8
xar   z4.h, z4.h, z5.h, #2
xar   z6.h, z6.h, z7.h, #16
xar   z8.s, z8.s, z9.s, #3
xar   z10.s, z10.s, z11.s, #32
xar   z12.d, z12.d, z13.d, #4
xar   z14.d, z14.d, z15.d, #64
index z0.b, #-0x10, #0xF
index z1.h, #0xF, #-0x10
index z2.s, #0, #0
index z3.d, #-5, #5
index z0.b, #-0x10, w0
index z1.h, #0, w1
index z2.s, #5, w2
index z3.d, #10, x3
index z4.b, #-0x10, wzr
index z5.d, #15, xzr
index z0.b, w0, #-0x10
index z1.h, w1, #0
index z2.s, w2, #5
index z3.d, x3, #10
index z4.b, wzr, #-0x10
index z5.d, xzr, #15
decb  x0, pow2
decd  x1, vl16, mul #3
dech  x2, vl32, mul #5
decw  x3, vl64, mul #7
incb  x4, vl128, mul #9
incd  x5, mul3, mul #10
inch  x6, mul4, mul #13
incw  x7, all, mul #16
decd  z0.d, pow2
dech  z1.h, vl2, mul #2
decw  z2.s, vl3, mul #4
incd  z3.d, vl4, mul #8
inch  z4.h, vl5, mul #12
incw  z5.s, vl6, mul #16
```

JitDisasm output:
```
xar     z0.b, z0.b, z1.b, #1
xar     z2.b, z2.b, z3.b, #8
xar     z4.h, z4.h, z5.h, #2
xar     z6.h, z6.h, z7.h, #16
xar     z8.s, z8.s, z9.s, #3
xar     z10.s, z10.s, z11.s, #32
xar     z12.d, z12.d, z13.d, #4
xar     z14.d, z14.d, z15.d, #64
index   z0.b, #-16, #15
index   z1.h, #15, #-16
index   z2.s, #0, #0
index   z3.d, #-5, #5
index   z0.b, #-16, w0
index   z1.h, #0, w1
index   z2.s, #5, w2
index   z3.d, #10, x3
index   z4.b, #-16, wzr
index   z5.d, #15, xzr
index   z0.b, w0, #-16
index   z1.h, w1, #0
index   z2.s, w2, #5
index   z3.d, x3, #10
index   z4.b, wzr, #-16
index   z5.d, xzr, #15
decb    x0, pow2
decd    x1, vl16, mul #3
dech    x2, vl32, mul #5
decw    x3, vl64, mul #7
incb    x4, vl128, mul #9
incd    x5, mul3, mul #10
inch    x6, mul4, mul #13
incw    x7, all, mul #16
decd    z0.d, pow2
dech    z1.h, vl2, mul #2
decw    z2.s, vl3, mul #4
incd    z3.d, vl4, mul #8
inch    z4.h, vl5, mul #12
incw    z5.s, vl6, mul #16
```

cc @dotnet/arm64-contrib

### CODE REVIEW

File: src/coreclr/jit/emitarm64.cpp [hunk 1]
```diff
@@ -9300,6 +9378,7 @@ void emitter::emitIns_R_I_I(instruction ins,
 
     id->idIns(ins);
     id->idInsFmt(fmt);
+    id->idInsOpt(opt);
```
Comment (TIHan) [inline src/coreclr/jit/emitarm64.cpp:9381]: Interesting that we weren't setting the `opt` before. Makes sense that we need to. I assume it doesn't cause any problems for the existing instructions.

File: src/coreclr/jit/emitarm64.cpp [hunk 2]
```diff
@@ -24094,8 +24345,9 @@ BYTE* emitter::emitOutput_InstrSve(BYTE* dst, instrDesc* id)
             dst += emitOutput_Instr(dst, code);
             break;
 
-        // Immediate and patterm to general purpose.
```
Comment (TIHan) [inline src/coreclr/jit/emitarm64.cpp:24097]: "patterm" xD

File: src/coreclr/jit/emitarm64.cpp [hunk 3]
```diff
@@ -24628,7 +24935,7 @@ BYTE* emitter::emitOutput_InstrSve(BYTE* dst, instrDesc* id)
             code = emitInsCodeSve(ins, fmt);
             code |= insEncodeReg_V_4_to_0(id->idReg1());                                           // ddddd
             code |= insEncodeReg_V_9_to_5(id->idReg2());                                           // nnnnn
-            code |= insEncodeSveElemsize_tszh_22_tszl_20_to_19(optGetSveElemsize(id->idInsOpt())); // xx
+            code |= insEncodeSveElemsize_tszh_23_tszl_20_to_19(optGetSveElemsize(id->idInsOpt())); // xx
```
Comment (TIHan) [inline src/coreclr/jit/emitarm64.cpp:24938]: The existing uses for `insEncodeSveElemsize_tszh_22_tszl_20_to_19` only assumed 1/2/4 byte sizes and not 8, so we should probably put an assert, before it `assert(optGetSveElemsize(id->idInsOpt()) == EA_8BYTE)`, **or** we could just have two functions for `tszh:tszl` with one that allows 1/2/4 byte and the other 1/2/4/8 byte.

File: src/coreclr/jit/emitarm64.cpp [hunk 4]
```diff
@@ -24638,7 +24945,7 @@ BYTE* emitter::emitOutput_InstrSve(BYTE* dst, instrDesc* id)
             code |= insEncodeReg_V_4_to_0(id->idReg1());                                           // ddddd
             code |= insEncodeReg_V_9_to_5(id->idReg2());                                           // nnnnn
             code |= insEncodeUimm5_20_to_16(emitGetInsSC(id));                                     // iii
-            code |= insEncodeSveElemsize_tszh_22_tszl_20_to_19(optGetSveElemsize(id->idInsOpt())); // xx
+            code |= insEncodeSveElemsize_tszh_23_tszl_20_to_19(optGetSveElemsize(id->idInsOpt())); // xx
```
Comment (TIHan) [inline src/coreclr/jit/emitarm64.cpp:24948]: Same as comment above regarding `tszh:tszl`.

File: src/coreclr/jit/emitarm64.cpp [hunk 5]
```diff
@@ -24648,7 +24955,7 @@ BYTE* emitter::emitOutput_InstrSve(BYTE* dst, instrDesc* id)
             code |= insEncodeReg_V_4_to_0(id->idReg1());                                           // ddddd
             code |= insEncodeReg_V_9_to_5(id->idReg2());                                           // nnnnn
             code |= insEncodeUimm5_20_to_16(insGetImmDiff(emitGetInsSC(id), id->idInsOpt()));      // iii
-            code |= insEncodeSveElemsize_tszh_22_tszl_20_to_19(optGetSveElemsize(id->idInsOpt())); // xx
+            code |= insEncodeSveElemsize_tszh_23_tszl_20_to_19(optGetSveElemsize(id->idInsOpt())); // xx
```
Comment (TIHan) [inline src/coreclr/jit/emitarm64.cpp:24958]: Same as comment above regarding `tszh:tszl`.

File: src/coreclr/jit/emitarm64.cpp [hunk 6]
```diff
@@ -21738,6 +21922,10 @@ void emitter::emitIns_Call(EmitCallType          callType,
             assert(isValidUimm5From1(imm));
             return (32 - imm);
 
+        case INS_OPTS_SCALABLE_D:
```
Comment (TIHan) [inline src/coreclr/jit/emitarm64.cpp:21925]: Do the existing uses of `insGetImmDiff` assume only B, H, S? If so, we should either put asserts before calling `insGetImmDiff` ensuring that we don't accidently pass D, **or** just have two different functions. This is a similar suggestion to `insEncodeSveElemsize_tszh_22_tszl_20_to_19` that I had.

**Review by TIHan — APPROVED** [review]
LGTM, just some suggestions for `insGetImmDiff` and `insEncodeSveElemsize_tszh_23_tszl_20_to_19` if they are applicable.

File: src/coreclr/jit/emitarm64.cpp (hunk 1 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/emitarm64.cpp:9381]: Yeah, the existing instructions don't seem to check `opt` anywhere else -- not even to assert that it's `INS_OPTS_NONE`. That's the default value in `emitIns_R_I_I`, so I imagine always initializing `idInsOpt` is fine; I didn't notice any issues.

File: src/coreclr/jit/emitarm64.cpp (hunk 3 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/emitarm64.cpp:24938]: I prefer the assert approach; I'll add those in.

Comment (a74nh) [reply src/coreclr/jit/emitarm64.cpp:24938]: +1 for asserts. The size restriction is (usually) a property of the instruction operation, whereas the helper function only cares about bit encodings.

**Review by a74nh — APPROVED** [review]
Assuming all the other review comments are fixed up, LGTM.

**Comment by amanasifkhalid** [comment]
Thanks for the reviews!

**Pull request merged**